File: src/backend/projetos/src/model/projeto_custo_model.py
from src import db
from src.model.fin_natureza_financeira_model import FinNaturezaFinanceiraModel


class ProjetoCustoModel(db.Model):
    __tablename__ = 'projeto_custo'

    id = db.Column(db.Integer, primary_key=True)
    nome = db.Column(db.String(100))
    valor_mensal = db.Column(db.Float)
    valor_total = db.Column(db.Float)
    justificativa = db.Column(db.Text)
    id_projeto_principal = db.Column(db.Integer, db.ForeignKey('projeto_principal.id'))
    id_fin_natureza_financeira = db.Column(db.Integer, db.ForeignKey('fin_natureza_financeira.id'))

    fin_natureza_financeira_model = db.relationship('FinNaturezaFinanceiraModel', foreign_keys=[id_fin_natureza_financeira])

    def mapping(self, data):
        self.id = data.get('id')
        self.id_projeto_principal = data.get('idProjetoPrincipal')
        self.id_fin_natureza_financeira = data.get('idFinNaturezaFinanceira')
        self.nome = data.get('nome')
        self.valor_mensal = data.get('valorMensal')
        self.valor_total = data.get('valorTotal')
        self.justificativa = data.get('justificativa')

    def serialize(self):
        return {
            'id': self.id,
            'idProjetoPrincipal': self.id_projeto_principal,
            'idFinNaturezaFinanceira': self.id_fin_natureza_financeira,
            'nome': self.nome,
            'valorMensal': self.valor_mensal,
            'valorTotal': self.valor_total,
            'justificativa': self.justificativa,
            'finNaturezaFinanceiraModel': self.fin_natureza_financeira_model.serialize() if self.fin_natureza_financeira_model else None,
        }

File: frontend/projetos/lib/app/data/model/projeto_custo_model.dart
import 'dart:convert';
import 'package:pluto_grid/pluto_grid.dart';

import 'package:projetos/app/data/model/model_imports.dart';

class ProjetoCustoModel {
	int? id;
	int? idProjetoPrincipal;
	int? idFinNaturezaFinanceira;
	String? nome;
	double? valorMensal;
	double? valorTotal;
	String? justificativa;
	FinNaturezaFinanceiraModel? finNaturezaFinanceiraModel;

	ProjetoCustoModel({
		this.id,
		this.idProjetoPrincipal,
		this.idFinNaturezaFinanceira,
		this.nome,
		this.valorMensal,
		this.valorTotal,
		this.justificativa,
		this.finNaturezaFinanceiraModel,
	});

	static List<String> dbColumns = <String>[
		'id',
		'nome',
		'valor_mensal',
		'valor_total',
		'justificativa',
	];
	
	static List<String> aliasColumns = <String>[
		'Id',
		'Nome',
		'Valor Mensal',
		'Valor Total',
		'Justificativa',
	];

	ProjetoCustoModel.fromJson(Map<String, dynamic> jsonData) {
		id = jsonData['id'];
		idProjetoPrincipal = jsonData['idProjetoPrincipal'];
		idFinNaturezaFinanceira = jsonData['idFinNaturezaFinanceira'];
		nome = jsonData['nome'];
		valorMensal = jsonData['valorMensal']?.toDouble();
		valorTotal = jsonData['valorTotal']?.toDouble();
		justificativa = jsonData['justificativa'];
		finNaturezaFinanceiraModel = jsonData['finNaturezaFinanceiraModel'] == null ? FinNaturezaFinanceiraModel() : FinNaturezaFinanceiraModel.fromJson(jsonData['finNaturezaFinanceiraModel']);
	}

	Map<String, dynamic> get toJson {
		Map<String, dynamic> jsonData = <String, dynamic>{};

		jsonData['id'] = id != 0 ? id : null;
		jsonData['idProjetoPrincipal'] = idProjetoPrincipal != 0 ? idProjetoPrincipal : null;
		jsonData['idFinNaturezaFinanceira'] = idFinNaturezaFinanceira != 0 ? idFinNaturezaFinanceira : null;
		jsonData['nome'] = nome;
		jsonData['valorMensal'] = valorMensal;
		jsonData['valorTotal'] = valorTotal;
		jsonData['justificativa'] = justificativa;
		jsonData['finNaturezaFinanceiraModel'] = finNaturezaFinanceiraModel?.toJson;
	
		return jsonData;
	}
	
	String objectEncodeJson() {
		final jsonData = toJson;
		return json.encode(jsonData);
	}

	plutoRowToObject(PlutoRow plutoRow) {
		id = plutoRow.cells['id']?.value;
		idProjetoPrincipal = plutoRow.cells['idProjetoPrincipal']?.value;
		idFinNaturezaFinanceira = plutoRow.cells['idFinNaturezaFinanceira']?.value;
		nome = plutoRow.cells['nome']?.value;
		valorMensal = plutoRow.cells['valorMensal']?.value?.toDouble();
		valorTotal = plutoRow.cells['valorTotal']?.value?.toDouble();
		justificativa = plutoRow.cells['justificativa']?.value;
		finNaturezaFinanceiraModel = FinNaturezaFinanceiraModel();
		finNaturezaFinanceiraModel?.descricao = plutoRow.cells['finNaturezaFinanceiraModel']?.value;
	}	

	ProjetoCustoModel clone() {
		return ProjetoCustoModel(
			id: id,
			idProjetoPrincipal: idProjetoPrincipal,
			idFinNaturezaFinanceira: idFinNaturezaFinanceira,
			nome: nome,
			valorMensal: valorMensal,
			valorTotal: valorTotal,
			justificativa: justificativa,
		);			
	}

	
}